deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AutoPromoter.java

```diff
@@ -122,85 +193,199 @@ boolean isCurrentDeployRetirable(String envId, DeployBean currDeployBean) throws
+            currEnvBean.getEnv_name());
+        if (result.getResult() == PromoteResult.ResultCode.PromoteBuild &&
+            StringUtils.isNotEmpty(result.getPromotedBuild())) {
+            safePromote(null, result.getPromotedBuild(), Constants.BUILD_STAGE, currDeployBean,
+                currEnvBean);
+        }
+    }
+
+    //This contains the logic about if there should be a promote deploy from the preceded
+    // environment.
+    public PromoteResult computePromoteDeployResult(EnvironBean currEnvBean,
+                                                    DeployBean currDeployBean,
+                                                    int size, PromoteBean promoteBean)
+        throws Exception {
+        String precededStage = promoteBean.getPred_stage();
+        // Special case when there is no preceded environment
+        EnvironBean
+            precededEnvBean =
+            environDAO.getByStage(currEnvBean.getEnv_name(), precededStage);
+        if (precededEnvBean == null) {
+            LOG.warn("Pred env {}/{} does not exist, bail out!", currEnvBean.getEnv_name(),
+                precededStage);
+            return new PromoteResult()
+                .withResultCode(PromoteResult.ResultCode.NoPredEnvironment);
         }
 
-        String predDeployId = predEnvBean.getDeploy_id();
+        String predDeployId = precededEnvBean.getDeploy_id();
         if (predDeployId == null) {
-            LOG.debug("Pred env {}/{} does not have deploy yet, bail out!", currEnvBean.getEnv_name(), predStage);
-            return;
+            LOG.debug("Pred env {}/{} does not have deploy yet, bail out!",
+                currEnvBean.getEnv_name(), precededStage);
+            return new PromoteResult()
+                .withResultCode(PromoteResult.ResultCode.NoPredEnvironmentDeploy);
         }
 
-        long currentDeployDate = 0;
-        if (currDeployBean != null) {
-            String fromDeployId = currDeployBean.getFrom_deploy();
-            if (fromDeployId != null) {
-                DeployBean fromDeployBean = deployDAO.getById(fromDeployId);
-                if (fromDeployBean.getEnv_id().equals(predEnvBean.getEnv_id())) {
-                    currentDeployDate = fromDeployBean.getStart_date();
-                } else {
-                    LOG.info("Current deploy {} in env {} was not promoted from pred env {}, but from env {}! Use the current deploy startDate",
-                        currDeployBean.getDeploy_id(), currEnvBean.getEnv_id(), predEnvBean.getEnv_id(), fromDeployBean.getEnv_id());
-                    currentDeployDate = currDeployBean.getStart_date();
-                }
-            } else {
-                LOG.info("Current deploy {} in env {} was not promoted from anywhere! Use the current deploy startDate",
-                    currDeployBean.getDeploy_id(), currEnvBean.getEnv_id());
-                currentDeployDate = currDeployBean.getStart_date();
+        Date now = new Date();
```

Comment:
This code from line 289 - 305 can be combined with similar ones in computePromoteBuildResult method. Create a new method?